presto-main/src/test/java/com/facebook/presto/util/TestFailures.java

```diff
@@ -0,0 +1,34 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.util;
+
+import com.google.common.collect.ImmutableList;
+import org.testng.annotations.Test;
+
+import static com.facebook.presto.util.Failures.toFailure;
+import static com.facebook.presto.util.Failures.toFailures;
+
+public class TestFailures
+{
+    @Test
+    public void testToFailureLoop()
+    {
+        Throwable exception1 = new RuntimeException("fake exception 1");
+        Throwable exception2 = new RuntimeException("fake exception 2");
+        exception1.addSuppressed(exception2);
+        exception2.addSuppressed(exception1);
```

Comment:
uh, but you should not do that, ever, should you?

Reply:
I accidentally did something to this extent (in an in-house connector). And the fact that `toFailure` doesn't handle that made the bug much worse.

Reply:
Yea... I don't object to `toFailure` safeguarding against cyclic suppression chain. 

However, i wonder -- is it possible to get such a case in current code base?

Reply:
Probably not in the open-source version. But good to have to prevent future production issues:
```
java.lang.StackOverflowError
        at com.facebook.presto.util.Failures.toFailure(Failures.java:68)
        at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193)
        at java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:948)
        at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481)
        at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
        at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:708)
        at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
        at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:499)
        at com.facebook.presto.util.Failures.toFailures(Failures.java:95)
        at com.facebook.presto.util.Failures.toFailure(Failures.java:77)
        at com.facebook.presto.util.Failures.toFailure(Failures.java:76)
        at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193)
        at java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:948)
        at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481)
        at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
        at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:708)
        at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
        at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:499)
        at com.facebook.presto.util.Failures.toFailures(Failures.java:95)
        at com.facebook.presto.util.Failures.toFailure(Failures.java:77)
        at com.facebook.presto.util.Failures.toFailure(Failures.java:76)
        at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193)
        at java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:948)
        at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481)
        at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
        at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:708)
        at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
        at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:499)
        at com.facebook.presto.util.Failures.toFailures(Failures.java:95)
        at com.facebook.presto.util.Failures.toFailure(Failures.java:77)
        at com.facebook.presto.util.Failures.toFailure(Failures.java:76)
        at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193)
        at java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:948)
        at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481)
        at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
        at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:708)
        at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
        at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:499)
...

Reply:
@findepi It can definitely happen when some one introduced a bug. And I think it's fair to have some protection because `toFailure` that throws makes it really hard to find the real problem.

Reply:
@haozhun i can only continue to agree :)